### PR TITLE
Unified routers

### PR DESCRIPTION
**Pull request by DZGoldman**
Unified ERC20 / Eth routers; in this flow, multiple sources sending funds to the same final destination can use the same router contracts (also no explicit on chain "allowance" step).

ParentChildRouter accepts native funds and any number of tokens; each ChildToParentRouter accepts native funds and one token (set on deployment).

### CODE REVIEW

File: script/FeeRouterDeployer/NovaToDAOTreasury/nova-deploy.s.sol [hunk 1]
```diff
@@ -6,10 +6,13 @@ contract DeployScript is Script {
         // deployed via eth-deploy.s.sol:
         address parentChainTarget = vm.envAddress("L1_TO_DAO_TIMELOCK_ROUTER");
```
Comment (godzillaba) [inline script/FeeRouterDeployer/NovaToDAOTreasury/nova-deploy.s.sol:7]: figure out what this is

**Comment by DZGoldman** [comment]
Testnet deployments: 
sepolia: 0x5DB54b55FfD0887Bf1634a865DB1e7db7a3dd797
arb-sepolia: 0xbDa3ECac5a9a7e61a69252648cC9C66c615ceAA9

File: src/FeeRouter/ParentToChildRewardRouter.sol [hunk 2]
```diff
@@ -93,4 +112,59 @@ contract ParentToChildRewardRouter is DistributionInterval {
         });
         emit FundsRouted(amount);
     }
+
+    /// @notice send full token balance in this contract to destination. Uses sender's address for fee refund
+    /// @param parentChainTokenAddr todor
```
Comment (gvladika) [inline src/FeeRouter/ParentToChildRewardRouter.sol:117]: comment needs update

File: src/FeeRouter/ParentToChildRewardRouter.sol [hunk 3]
```diff
@@ -93,4 +112,59 @@ contract ParentToChildRewardRouter is DistributionInterval {
...
+            _token: parentChainTokenAddr,
+            _refundTo: msg.sender, // send excess fees to the sender's address on the child chain
+            _to: destination,
+            _amount: amount,
+            _maxGas: gasLimit,
+            _gasPriceBid: maxFeePerGas,
+            _data: _data
+        });
+
+        emit FundsRouted(amount);
```
Comment (gvladika) [inline src/FeeRouter/ParentToChildRewardRouter.sol:168]: It's the same event that gets emitted when routing Eth. Something like `event FundsRouted(address indexed token, uint256 amount)` would be more useful (also in the case of ChildToParent router)

File: src/FeeRouter/ChildToParentRewardRouter.sol [hunk 4]
```diff
@@ -2,39 +2,86 @@
...
-            _updateDistribution();
+        if (canDistribute(NATIVE_CURRENCY) && value > 0) {
+            _updateDistribution(NATIVE_CURRENCY);
             IArbSys(address(100)).withdrawEth{value: value}(parentChainTarget);
             emit FundsRouted(value);
         }
     }
+
+    /// @notice send all of provided token in this contract to destination on parent chain
+    /// @param _parentChainTokenAddr parent chain (i.e., chaun underlying this one) address of token to route
```
[the remark this answers is not in the document]
Comment (gzeoneth) [reply src/FeeRouter/ChildToParentRewardRouter.sol:64]: actually seems like we need to change this function, let's discsuss internally

File: src/FeeRouter/ChildToParentRewardRouter.sol [hunk 5]
```diff
@@ -2,39 +2,86 @@
...
+            revert DistributionTooSoon(block.timestamp, nextDistributions[_parentChainTokenAddr]);
+        }
+        if (value == 0) {
+            revert NoValue(_parentChainTokenAddr);
+        }
+        IERC20(childChainTokenAddress).approve(gateway, value);
+
+        childChainGatewayRouter.outboundTransfer(_parentChainTokenAddr, parentChainTarget, value, "");
+
+        _updateDistribution(_parentChainTokenAddr);
```
[the remark this answers is not in the document]
Comment (gzeoneth) [reply src/FeeRouter/ChildToParentRewardRouter.sol:85]: agree

**Review by gzeoneth — CHANGES_REQUESTED** [review]

**Pull request merged**